src/libcore/iter.rs

```diff
@@ -115,7 +119,10 @@ pub trait Iterator {
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
     fn count(self) -> usize where Self: Sized {
-        self.fold(0, |cnt, _x| cnt + 1)
+        self.fold(0, |cnt, _| match cnt.checked_add(1) {
+            Some(c) => c,
+            None => panic!("overflow while counting the elements of an iterator"),
```

Comment:
Due to the way we monomorphize/translate this function it turns out that a panic will happen if it's included in an unoptimized build, and I'm not sure that we can claim that this will *never* be used in a hot spot, so I would be wary of taking the perf hit for something that will only incredibly rarely get hit.